docs/development/etcdoperatortask-design.md

```diff
@@ -0,0 +1,481 @@
+    // Add config fields for other tasks
+}
+
+type OnDemandSnapshotConfig struct {
+    // +required
+    // +kubebuilder:validation:Enum=full;delta
+    SnapshotType    *string
+    // +optional
+    // +kubebuilder:default:=60
+    // +kubenuilder:validation:XValidation:message="TTLSecondsAfterFinished must be greater than 0",rule="!has(self) || self > 0"
+    TimeoutSeconds  *int 
+}
+```
+
+> [!Note]
+> Use `runtime.RawExtension` for the `config` field to enable each task type to define its own schema dynamically. Ensure this field is validated through an admission webhook to enforce type safety and provide immediate feedback to users.
+
+### Status Subresource
+
+The `status` field tracks the progress and outcome of each task, including state transitions, errors, and operation history.
+
+
+
+```go
+// EtcdOperatorTaskStatus is the status for a EtcdOperatorTask resource.
+type EtcdOperatorTaskStatus struct {
+  // State is the last known state of the task.
+  State TaskState `json:"state"`
+  // Initiated at records the time at which the task has moved from "pending" state to the "InProgress" state.
+  // +optional
+  InitiatedAt *metav1.Time `json:"initiatedAt,omitempty"`
+  // LastErrors represents the errors when processing the task. Will have a limit of 10 entries at a time.
+  // +optional
+  LastErrors []LastError `json:"lastErrors,omitempty"`
+  // LastOperation captures the last operation status if task involves many stages.
+  // +optional
+  LastOperation *LastOperation `json:"lastOperation,omitempty"`
+}
+
+type LastOperation struct {
+  // Status of the last operation, one of pending, progress, completed, failed.
+  State OperationState `json:"state"`
+  // Phase represents the current phase of the operation wrt the interface methods
+  Phase OperationPhase `json:"phase"`
+  // LastTransitionTime records the timestamp of the most recent state transition, marking when the operation moved from its previous state to the current value specified in the 'State' field.
+  LastTransitionTime *metav1.Time `json:"lastTransitionTime"`
+  // A human readable message indicating details about the last operation.
+  Description string `json:"description"`
+}
+
+// LastError stores details of the most recent error encountered for the task.
+type LastError struct {
+  // Code is an error code that uniquely identifies an error.
+  Code ErrorCode `json:"code"`
+  // Description is a human-readable message indicating details of the error.
+  Description string `json:"description"`
+  // ObservedAt is the time at which the error was observed.
+  ObservedAt *metav1.Time `json:"observedAt"`
+}
+
+// TaskState represents the state of the task ie the status of the CR.
+type TaskState string
+
+const (
+    TaskStatePending TaskState = "Pending"
+    TaskStateRejected TaskState = "Rejected"
+    TaskStateInProgress TaskState = "InProgress"
+    TaskStateFailed TaskState = "Failed"
+    TaskStateSucceeded TaskState = "Succeeded"
+)
```

Comment:
The implementation behavior below assumes the state transitions are strict across tasks. 
Should there be a state diagram here to ensure the implementor of future tasks understand what state a task can go so that they abide by the flow and map the behavior accordingly in their implementation.